### PR TITLE
fix: lower string literals with actual value instead of default

### LINKED ISSUE #11582: Value of string constant is not displayed correctly when hovered over
**rust-analyzer version**: `rust-analyzer 4e72700e3 2022-02-28 stable`

**rustc version**: `rustc 1.59.0 (9d1b2106e 2022-02-23)`

**relevant settings**: None

The definition incorrectly shows that `HELLO_WORLD` is an empty string.

### PR DESCRIPTION
**Pull request by tysg**
Fixes #11582. Some questions below in the code review section.

### CODE REVIEW

File: crates/hir_def/src/body/lower.rs [hunk 1]
```diff
@@ -923,7 +923,10 @@ impl From<ast::LiteralKind> for Literal {
                 let text = bs.value().map(Box::from).unwrap_or_else(Default::default);
                 Literal::ByteString(text)
             }
-            LiteralKind::String(_) => Literal::String(Default::default()),
+            LiteralKind::String(str) => {
+                let text = str.value().map(Box::from).unwrap_or_else(Default::default);
+                Literal::String(text)
+            }
             LiteralKind::Byte => Literal::Uint(Default::default(), Some(BuiltinUint::U8)),
```
Comment (tysg) [inline crates/hir_def/src/body/lower.rs:930]: In L905, there's a FIXME item:
```
// FIXME: these should have actual values filled in, but unsure on perf impact
```

1) What's a good way to measure the performance impact? I believe it's mostly changes in memory consumption, but I can only find metrics on computational performance/speed (via `integrated_benchmarks::integrated_highlighting_benchmark`)
2) Should we apply similar changes to the rest of the `LiteralKind` enums (i.e. `Byte` and `Char`)

Comment (lnicola) [reply crates/hir_def/src/body/lower.rs:930]: `$ rust-analyzer analysis-stats --memory-usage .` might work.

Comment (lnicola) [reply crates/hir_def/src/body/lower.rs:930]: 2. That sounds reasonable, and we might need them for const-eval at some point.

**Comment by tysg** [comment]
Did some memory profiling using `cargo run --bin rust-analyzer --release -- -q analysis-stats --memory-usage .`

Before (9d473a0b9f92c0739e4635fa8ef95a91ac2a481d):
```
Running `target/release/rust-analyzer -q analysis-stats --memory-usage .`
Failed to create perf counter: Operation not permitted (os error 1)
Failed to create perf counter: Operation not permitted (os error 1)
Database loaded:     392.17ms, 98mb (metadata 222.96ms, 489kb; build 68.66ms, 43kb)
Failed to create perf counter: Operation not permitted (os error 1)
  crates: 39, mods: 854, decls: 19304, fns: 14437
Item Collection:     8.08s, 352mb
Failed to create perf counter: Operation not permitted (os error 1)
  exprs: 397281, ??ty: 336 (0%), ?ty: 158 (0%), !ty: 144
Inference:           41.06s, 683mb
Total:               49.14s, 1035mb
```

After:
```
Failed to create perf counter: Operation not permitted (os error 1)
Failed to create perf counter: Operation not permitted (os error 1)
Database loaded:     9.38s, 98mb (metadata 224.45ms, 489kb; build 9.05s, 46kb)
Failed to create perf counter: Operation not permitted (os error 1)
  crates: 39, mods: 854, decls: 19304, fns: 14437
Item Collection:     8.01s, 352mb
Failed to create perf counter: Operation not permitted (os error 1)
  exprs: 397297, ??ty: 336 (0%), ?ty: 163 (0%), !ty: 144
Inference:           42.21s, 684mb
Total:               50.22s, 1037mb
```
Looks like there's no significant change in memory usage as I understand. 

Regarding `Byte` and `Char`, since the `LiteralKind` intermediate doesn't record the value of the Byte and Char literals, we can't fix it like this. I don't think using the `LiteralKind` intermediate is the best way to lower an AST to HIR (I asked [a question about this on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/lowering.3A.20ast.3A.3ALiteral.20-.3E.20LiteralKind.20-.3E.20hir.3A.3ALiteral)). Should the team think so too, I'll send in the changes in a separate PR.

So this PR is ready for merge/review now @lnicola, thanks!

**Review by lnicola — APPROVED** [review]

**Pull request merged**